internal/backend/oauth/webhook.go

```diff
@@ -1,134 +1,167 @@
+
+	ctx := r.Context()
+	i, err := s.svcs.Integrations().GetByName(ctx, intgSymbol)
+	if err != nil {
+		l.Warn("Failed to retrieve integration", zap.Error(err))
+		http.Error(w, "failed to retrieve integration", http.StatusInternalServerError)
+		return
+
+	}
+
+	conns, err := s.svcs.Connections().List(ctx, sdkservices.ListConnectionsFilter{
+		IntegrationID: i.Get().ID(),
+	})
+	if err != nil {
+		l.Warn("Failed to retrieve connections", zap.Error(err))
+		http.Error(w, "failed to retrieve connections", http.StatusInternalServerError)
+		return
+	}
+
+	// ensure the specified integration is registered.
+	o := s.svcs.OAuth()
+	if _, _, err := o.Get(ctx, r.PathValue("intg")); err != nil {
+		l.Warn("OAuth config retrieval error", zap.Error(err))
+		http.Error(w, "unrecognized integration name", http.StatusBadRequest)
 		return
 	}
 
 	// Read the temporary authorization code from the redirected request.
 	code := r.FormValue("code")
 	if code == "" {
-		s.logger.Warn("OAuth redirect request without code parameter",
-			zap.String("urlPath", r.URL.Path),
-			zap.Any("form", r.Form),
-		)
-		destURL += "&error=" + url.QueryEscape("Missing code parameter")
-		http.Redirect(w, r, destURL, http.StatusFound)
+		l.Warn("OAuth redirect request without code parameter")
+		http.Error(w, "missing code parameter", http.StatusBadRequest)
 		return
 	}
 
+	state := r.FormValue("state")
+
 	// Convert it into an OAuth refresh token / user access token.
-	token, err := o.Exchange(ctx, id, state, code)
+	token, err := o.Exchange(ctx, intg, state, code)
+	if err != nil {
+		l.Warn("OAuth exchange error", zap.Error(err))
+		http.Error(w, "exchange error", http.StatusBadRequest)
+		return
+	}
+
+	l.Info("OAuth token exchange successful")
+
+	oauthData, err := sdkintegrations.OAuthData{Token: token, Params: r.URL.Query()}.Encode()
 	if err != nil {
-		s.logger.Warn("OAuth exchange error",
-			zap.String("urlPath", r.URL.Path),
-			zap.Error(err),
-		)
-		destURL += "&error=" + url.QueryEscape(err.Error())
-		http.Redirect(w, r, destURL, http.StatusFound)
+		l.Warn("Failed to encode OAuth token", zap.Error(err))
+		http.Error(w, "failed to encode OAuth token", http.StatusInternalServerError)
 		return
 	}
 
-	// Pass the new OAuth token along with the redirected request's
-	// URL parameters to the initiator of the OAuth flow.
-	destURL += "&ak_token_access=" + url.QueryEscape(token.AccessToken)
-	destURL += "&ak_token_refresh=" + url.QueryEscape(token.RefreshToken)
-	destURL += "&ak_token_type=" + url.QueryEscape(token.TokenType)
-	destURL += "&ak_token_expiry=" + url.QueryEscape(token.Expiry.UTC().Format(time.RFC3339Nano))
-	destURL = strings.ReplaceAll(destURL, "?&", "?")
-	http.Redirect(w, r, destURL, http.StatusFound)
+	type conn struct {
+		ID          string
+		Name        string
+		ProjectName string
+	}
+
+	data := struct {
+		Intg      string
+		OAuthData string
+		Conns     []conn
+	}{
+		Intg:      intg,
+		OAuthData: oauthData,
+	}
```

Comment:
The `exchange()` is already too long. Please define structs outside it, even if they're used only here.

Reply:
tbh i disagree here, if this is the only place it is used, it needs to stay here.

Reply:
you know what, scratch that. will fix.

Reply:
looks like there was a lot of unused code here from experimentation i did prior. removed.